File: parsers/src/main/java/vct/parsers/ColCParser.java
package vct.parsers;

import static hre.lang.System.*;

import java.io.*;
import java.nio.file.Path;
import java.nio.file.Paths;

import hre.config.Configuration;
import hre.lang.HREExitException;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import scala.NotImplementedError;
import vct.col.ast.stmt.decl.ProgramUnit;

/**
 * Parse specified code and convert the contents to COL.
 */
public class ColCParser extends ColIParser {

    @Override
    public ProgramUnit parse(CharStream input, String file_name) {
        throw new NotImplementedError(
                "Cannot parse a C file from a CharStream: " +
                "the preprocessor requires a byte stream.");
    }

    @Override
    public ProgramUnit parse(InputStream stream, String file_name) {
        try {
            Runtime runtime = Runtime.getRuntime();

            StringBuilder command = new StringBuilder(Configuration.cpp_command.get());
            command.append(" -nostdinc -nocudainc -nocudalib -isystem ").append(Configuration.getCIncludePath().getAbsolutePath());

            Path filePath = Paths.get(file_name).toAbsolutePath().getParent();

            if(filePath != null) {
                // Not quite correct, because this allows <> includes to see local files as well as "" includes.
                command.append(" -I").append(filePath.toString());
            }

            for (String p : Configuration.cpp_include_path) {
                command.append(" -I").append(p);
            }
            for (String p : Configuration.cpp_defines) {
                command.append(" -D").append(p);
            }
            command.append(" -");

            Progress("pre-processing command line: %s", command.toString());

            final Process process = runtime.exec(command.toString());

            new Thread(() -> {
                try {
                    byte[] buf = new byte[4096];
                    int read = stream.read(buf);

                    while(read != -1) {
                        process.getOutputStream().write(buf, 0, read);
                        read = stream.read(buf);
                    }

                    process.getOutputStream().close();
                } catch(IOException e) {
                    DebugException(e);
                }
            }).start();

            Thread t = new Thread(() -> {
                BufferedReader err = new BufferedReader(new InputStreamReader(process.getErrorStream()));
                String s;
                try {
                    while ((s = err.readLine()) != null) {
                        Output("%s", s);
                    }
                } catch (IOException e) {
                    DebugException(e);
                }
            });
            t.setDaemon(true);
            t.start();
            return super.parse(CharStreams.fromStream(process.getInputStream()), file_name);
        } catch (Exception e) {
            DebugException(e);
            Abort("Exception %s while parsing %s", e.getClass(), file_name);
        }
        return null;
    }
}

File: parsers/src/main/java/vct/parsers/ColCParser.scala
package vct.parsers
import com.typesafe.scalalogging.LazyLogging
import hre.config.Configuration
import hre.io.{RWFile, Readable}
import org.antlr.v4.runtime.{CharStream, CharStreams}
import vct.parsers.CParser.PreprocessorError
import vct.parsers.transform.{BlameProvider, InterpretedFileOriginProvider, OriginProvider}
import vct.result.VerificationError.{Unreachable, UserError}

import java.io.{File, FileNotFoundException, InputStreamReader, OutputStreamWriter, StringWriter}
import java.nio.charset.StandardCharsets
import java.nio.file.{Path, Paths}

case object CParser {
  case class PreprocessorError(fileName: String, errorCode: Int, error: String) extends UserError {
    override def code: String = "preprocessorError"
    override def text: String =
      s"Preprocesing file $fileName failed with exit code $errorCode:\n$error"
  }
}

case class ColCParser(override val originProvider: OriginProvider,
                      override val blameProvider: BlameProvider,
                      cc: Path,
                      systemInclude: Path,
                      otherIncludes: Seq[Path],
                      defines: Map[String, String],
                      language: Language) extends Parser(originProvider, blameProvider) with LazyLogging {
  def interpret(localInclude: Seq[Path], input: String, output: String): Process = {
    var command = Seq(cc.toString, "-C", "-E")

    command ++= Seq("-nostdinc", "-nocudainc", "-nocudalib", "--cuda-host-only")
    command ++= Seq("-isystem", systemInclude.toAbsolutePath.toString)

    command ++= localInclude.map("-I" + _.toAbsolutePath)
    command ++= otherIncludes.map("-I" + _.toAbsolutePath.toString)
    command ++= defines.map { case (k, v) => s"-D$k=$v" }
    command ++= Seq("-o", output)
    command :+= input

    logger.debug(command.toString())

    new ProcessBuilder(command:_*).start()
  }

  override def parse[G](stream: CharStream): ParseResult[G] = {
    throw Unreachable("Should not parse C files from an ANTLR CharStream: they need to be interpreted first!")
  }

  override def parse[G](readable: Readable): ParseResult[G] =
    try {
      val interpreted = File.createTempFile("vercors-interpreted-", ".i")
      interpreted.deleteOnExit()

      val process = interpret(
        localInclude=Option(Paths.get(readable.fileName).getParent).toSeq,
        input="-",
        output=interpreted.toString
      )
      new Thread(() => {
        val writer = new OutputStreamWriter(process.getOutputStream, StandardCharsets.UTF_8)
        readable.read { reader =>
          val written = reader.transferTo(writer)
          logger.debug(s"Wrote $written bytes to clang")
          writer.close()
        }
      }).start()
      process.waitFor()

      if(process.exitValue() != 0) {
        val writer = new StringWriter()
        new InputStreamReader(process.getInputStream).transferTo(writer)
        new InputStreamReader(process.getErrorStream).transferTo(writer)
        writer.close()
        throw PreprocessorError(readable.fileName, process.exitValue(), writer.toString)
      }

      val result = ColIParser(InterpretedFileOriginProvider(originProvider, RWFile(interpreted)), blameProvider).parse[G](RWFile(interpreted))
      result
    } catch {
      case _: FileNotFoundException => throw FileNotFound(readable.fileName)
    }
}
